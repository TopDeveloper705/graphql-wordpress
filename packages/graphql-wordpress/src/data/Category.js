import { toGlobalId } from 'graphql-relay';

class Category {
  getID() {
    return toGlobalId(this.constructor.name, this.id);
  }

  static getEndpoint() {
    return process.env.WP_CATEGORIES_ENDPOINT || 'wp/v2/categories';
  }
}

export default Category;
