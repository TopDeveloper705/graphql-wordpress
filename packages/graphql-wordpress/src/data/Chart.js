import { toGlobalId } from 'graphql-relay';

class Chart {
  getID() {
    return toGlobalId(this.constructor.name, 'me');
  }

  static getEndpoint() {
    return 'https://itunes.apple.com/us/rss/topalbums/limit=25/explicit=true/json';
  }
}

export default Chart;
