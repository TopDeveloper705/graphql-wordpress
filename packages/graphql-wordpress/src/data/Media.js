import { toGlobalId } from 'graphql-relay';

class Media {
  getID() {
    return toGlobalId(this.constructor.name, this.id);
  }

  static getEndpoint() {
    return process.env.WP_MEDIA_ENDPOINT || 'wp/v2/media';
  }
}

export default Media;
