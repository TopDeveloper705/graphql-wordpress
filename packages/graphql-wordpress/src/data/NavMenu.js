import { toGlobalId } from 'graphql-relay';

class NavMenu {
  getID() {
    return toGlobalId(this.constructor.name, this.id);
  }

  static getEndpoint() {
    return process.env.WP_NAV_MENUS_ENDPOINT || 'graphql/v1/nav-menus';
  }
}

export default NavMenu;
