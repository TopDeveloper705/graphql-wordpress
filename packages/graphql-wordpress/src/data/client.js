import redis from 'redis';
import bluebird from 'bluebird';

/* eslint-disable no-console */

bluebird.promisifyAll(redis.RedisClient.prototype);
bluebird.promisifyAll(redis.Multi.prototype);

// This value needs to be the same on the Relay server
const REDIS_PREFIX = 'wpgql:';
export const HASH_KEY = 'queries';

let client;

export default function getClient() {
  if (client) {
    return client;
  }

  client = redis.createClient({ prefix: REDIS_PREFIX });
  // use this to debug Redis operations
  // client.monitor();
  client.on('error', err => {
    console.log(`Error ${err}`);
  });

  client.on('connect', () => {
    console.log('Redis client connected');
  });

  client.on('monitor', (time, args) => {
    console.log('\n', args);
  });

  return client;
}
