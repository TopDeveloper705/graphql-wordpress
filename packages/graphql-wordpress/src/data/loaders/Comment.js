import Dataloader from 'dataloader';
import fetchData from 'data/utils';
import Comment from 'data/Comment';

// Dataloader expects IDs that can be read by the REST API

export default function getCommentLoaders() {
  const endpoint = Comment.getEndpoint();

  const commentLoader = new Dataloader(ids =>
    fetchData(endpoint, {
      qs: { include: ids, orderby: 'include', per_page: 100 },
    }).then(({ data: { body } }) => body)
  );

  return {
    load: async id => {
      const data = await commentLoader.load(id);
      return data ? Object.assign(new Comment(), data) : null;
    },
  };
}
