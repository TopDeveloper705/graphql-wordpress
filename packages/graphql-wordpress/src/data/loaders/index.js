import getPostLoaders from './Post';
import getPageLoaders from './Page';
import getTaxonomyLoaders from './Taxonomy';
import getCategoryLoaders from './Category';
import getCommentLoaders from './Comment';
import getMediaLoaders from './Media';
import getNavMenuLoaders from './NavMenu';
import getUserLoaders from './User';
import getSettingsLoaders from './Settings';
import getSidebarLoaders from './Sidebar';
import getTagLoaders from './Tag';
import getChartLoaders from './Chart';

export default function getLoaders() {
  return {
    Post: getPostLoaders(),
    Page: getPageLoaders(),
    Taxonomy: getTaxonomyLoaders(),
    Category: getCategoryLoaders(),
    Comment: getCommentLoaders(),
    Media: getMediaLoaders(),
    NavMenu: getNavMenuLoaders(),
    User: getUserLoaders(),
    Settings: getSettingsLoaders(),
    Sidebar: getSidebarLoaders(),
    Tag: getTagLoaders(),
    Chart: getChartLoaders(),
  };
}
