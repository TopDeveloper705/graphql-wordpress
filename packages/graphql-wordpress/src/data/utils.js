import url from 'url';
import request from 'request-promise';
import getClient from 'data/client';

/* eslint-disable no-console */

export const toBase64 = str => new Buffer(str).toString('base64');

const rp = (path, opts = {}) => {
  let uri = path;
  const urlObj = url.parse(path);
  if (!urlObj.host) {
    uri = `${process.env.WP_API_HOST}${path}`;
  }

  const params = Object.assign(
    {
      uri,
      jar: true,
      json: true,
      simple: true,
      strictSSL: false,
    },
    opts
  );

  return request(params);
};

export const clearEndpointCache = path => {
  const client = getClient();
  // eslint-disable-next-line no-console
  console.log(`Clearing cache for: ${path}`);
  return new Promise((resolve, reject) => {
    client.hkeys(path, (err, keys) => {
      if (err) {
        console.log(`Cache lookup failed for: ${path}`);
        return reject(err);
      }
      return resolve(keys.map(hashKey => client.hdelAsync(path, hashKey)));
    });
  });
};

const fetchData = (path, opts = {}) => {
  const key = toBase64(`${JSON.stringify(opts)}`);
  const client = getClient();

  const params = {
    ...opts,
    resolveWithFullResponse: true,
  };

  return new Promise((resolve, reject) => {
    const method = params.method || 'GET';
    const makeRequest = (mutation = false) =>
      rp(path, params)
        .catch(error => {
          console.log(`${method} Request failed for: ${path}`);
          console.log(error);
          reject(error);
        })
        .then(async response => {
          if (response) {
            console.log(`Made ${method} request for: ${path}`);
            if (mutation) {
              await clearEndpointCache(path);
            }
            client.hset(path, key, JSON.stringify(response));
            resolve({
              cache: 'miss',
              data: response,
            });
          } else {
            reject();
          }
        });

    if (params.method && params.method !== 'GET') {
      makeRequest(true);
    } else {
      client.hget(path, key, (err, cached) => {
        if (err) {
          reject(err);
        } else if (cached) {
          console.log(`Resolving from cache: ${path}`);
          resolve({
            cache: 'hit',
            data: JSON.parse(cached),
          });
        } else {
          makeRequest();
        }
      });
    }
  });
};

export default fetchData;
