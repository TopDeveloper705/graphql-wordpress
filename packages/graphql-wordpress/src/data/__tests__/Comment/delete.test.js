import Comment from 'data/Comment';
import * as utils from 'data/utils';

utils.default = jest.fn(() => ({
  data: {
    body: {
      id: 13,
      status: 'trash',
    },
  },
}));

utils.clearEndpointCache = jest.fn(() => Promise.resolve());

describe('Test Comment CRUD: delete', () => {
  test('Test delete', async () => {
    const input = {
      id: 13,
    };
    const deletion = await Comment.delete(input);
    expect(deletion).toMatchSnapshot();
  });
});
