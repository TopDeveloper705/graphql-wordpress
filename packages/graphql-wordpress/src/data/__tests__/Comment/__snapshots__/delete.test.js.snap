// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Comment CRUD: delete Test delete 1`] = `
Object {
  "status": "delete",
}
`;
